### PR TITLE
Fix bug in sigma clipping return bounds

### LINKED ISSUE #11850: sigma_clip does not return a masked array when input is empty
### Description

When using sigma_clip() with default parameters (or when selecting masked=True) and passing an empty array to it, sigma_clip returns an empty np.array instead of an empty masked array. The result is that one cannot currently rely on sigma_clip to return a masked array.

### Expected behavior

when given [] as input, sigma_clip should return [] either as a masked_array or as a regular array, depending on the masked option that was used.

### Steps to Reproduce

1. [First Step]
2. [Second Step]
3. [and so on...]

```python
# Put your Python code snippet here.
```

### LINKED ISSUE #11992: SigmaClip to return NaN as bounds when all input array is masked
### Description

In the doc (e.g., [`sigma_clip`](https://docs.astropy.org/en/stable/api/astropy.stats.sigma_clip.html#astropy.stats.sigma_clip)), it is said that "the minimum and maximum clipping bounds are also returned" if `return_bounds=True`. I just happened to encounter a problem when I want the ``result`` AND the bounds ``low`` and ``upp``.

The simplest example to show this :
```python
import numpy as np
from astropy.stats import sigma_clip

a_allmask = np.ma.array(np.arange(10.), mask=[True]*10)
a = np.ma.array(np.arange(10.), mask=[False]+[True]*9)

for data in [a, a_allmask]:
    res, low, upp = sigma_clip(data, return_bounds=True)
    print(low, upp)
```
This raised ``ValueError``:
```
0.0 0.0
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-18-b2add72739a4> in <module>
      6
      7 for data in [a, a_allmask]:
----> 8     res, low, upp = sigma_clip(data, return_bounds=True)
      9     print(low, upp)

ValueError: too many values to unpack (expected 3)
```

So it means lower/upper bounds are not returned at all, even though `return_bounds=True`. Maybe better to let it return NaN for both lower/upper bounds if all data is masked for consistency.

### PR DESCRIPTION
**Pull request by larrybradley**
### Description

This PR fixes a bug in sigma clipping where the bounds would not be returned for completely empty or masked data.  Some documentation improvements and PEP8 fixes came along for the ride.

Fixes #11992
Fixes #11850

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label.
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

File: astropy/stats/sigma_clipping.py [hunk 1]
```diff
@@ -385,12 +395,12 @@ def _sigmaclip_fast(self, data, axis=None,
     def _sigmaclip_noaxis(self, data, masked=True, return_bounds=False,
                           copy=True):
         """
-        Sigma clip the data when ``axis`` is None and ``grow`` is not >0.
+        Sigma clip the data when ``axis`` is None and ``grow`` is not
```
Comment (mhvk) [inline astropy/stats/sigma_clipping.py:398]: Total nitpick, but my tendency would be to try to keep a single line and instead delete "the data" or so

File: astropy/stats/sigma_clipping.py [hunk 2]
```diff
@@ -231,12 +236,16 @@ def __init__(self, sigma=3., sigma_lower=None, sigma_upper=None,
...
+        self._niterations = 0
         self.grow = grow
 
-        # This just checks that SciPy is available, to avoid failing later
-        # than necessary if __call__ needs it:
+        # This just checks that SciPy is available, to avoid failing
+        # later than necessary if __call__ needs it:
         if self.grow:
             from scipy.ndimage import binary_dilation
+            self.binary_dilation = binary_dilation
```
Comment (mhvk) [inline astropy/stats/sigma_clipping.py:248]: Maybe make this a private attribute? (I thought about an try/except import on top which would set `binary_dilation = None` if scipy is not available, but think the import here, only when needed, is better, saving import time and memory for the more common case where `grow` is not used).

File: astropy/stats/sigma_clipping.py [hunk 3]
```diff
@@ -575,43 +587,52 @@ def __call__(self, data, axis=None, masked=True, return_bounds=False,
...
-            the input mask was `True`.  If ``return_bounds=True`` then
+            the input mask was `True`. If ``return_bounds=True`` then
             the returned minimum and maximum clipping thresholds will be
             be `~numpy.ndarray`\\s.
         """
-
         data = np.asanyarray(data)
 
         if data.size == 0:
-            return data
```
Comment (mhvk) [inline astropy/stats/sigma_clipping.py:596]: Oops!

File: astropy/stats/sigma_clipping.py [hunk 4]
```diff
@@ -575,43 +587,52 @@ def __call__(self, data, axis=None, masked=True, return_bounds=False,
...
+                return data, self._min_value, self._max_value
+            else:
+                return data
 
         if isinstance(data, np.ma.MaskedArray) and data.mask.all():
             if masked:
-                return data
+                result = data
+            else:
+                result = np.ma.filled(data.astype(float), fill_value=np.nan)
```
Comment (mhvk) [inline astropy/stats/sigma_clipping.py:616]: Maybe just `np.full(data.shape, np.nan)`?

File: astropy/stats/sigma_clipping.py [hunk 5]
```diff
@@ -634,24 +655,29 @@ def sigma_clip(data, sigma=3, sigma_lower=None, sigma_upper=None, maxiters=5,
...
+        `scipy.stats.sigmaclip` provides a subset of the functionality
+        in this class. Also, its input data cannot be a masked array
+        and it does not handle data that contains invalid values (i.e.,
+        NaN or inf). Also note that it uses the mean as the centering
+        function.
+
+        If your data is a `~numpy.ndarray` with no invalid values
+        and you want to use the mean as the centering function
+        with ``axis=None`` and iterate to convergence, then
         `scipy.stats.sigmaclip` is ~25-30% faster than the equivalent
```
Comment (mhvk) [inline astropy/stats/sigma_clipping.py:681]: Is this still true with the new gufunc?

File: astropy/stats/tests/test_sigma_clipping.py [hunk 6]
```diff
@@ -239,30 +239,45 @@ def test_sigmaclip_fully_masked():
...
+    clipped_data, low, high = sigma_clip(data, return_bounds=True)
+    assert np.ma.allequal(data, clipped_data)
+    assert np.isnan(low)
+    assert np.isnan(high)
+
 
 def test_sigmaclip_empty_masked():
-    """Make sure a empty masked array is returned when sigma clipping an empty
+    """
+    Make sure a empty masked array is returned when sigma clipping an empty
```
Comment (mhvk) [inline astropy/stats/tests/test_sigma_clipping.py:250]: a -> an empty

File: astropy/stats/tests/test_sigma_clipping.py [hunk 7]
```diff
@@ -239,30 +239,45 @@ def test_sigmaclip_fully_masked():
...
+    assert np.ma.allequal(data, clipped_data)
+    assert np.isnan(low)
+    assert np.isnan(high)
+
 
 def test_sigmaclip_empty():
-    """Make sure a empty array is returned when sigma clipping an empty array.
     """
-
+    Make sure a empty array is returned when sigma clipping an empty
```
Comment (mhvk) [inline astropy/stats/tests/test_sigma_clipping.py:265]: a -> an

**Review by mhvk — COMMENTED** [review]
Looks good, with only some very nitpicky comments. I think it might have been easier to keep the docs cleanup separate from the bug fix, but since this is just going back to 4.3.2, it should not be a problem to backport.

**Review by pllim — COMMENTED** [review]
I have nothing to add to review from @mhvk . Thanks for the fix!

File: astropy/stats/sigma_clipping.py (hunk 1 above)
Comment (larrybradley) [reply astropy/stats/sigma_clipping.py]: fixed

File: astropy/stats/sigma_clipping.py (hunk 2 above)
Comment (larrybradley) [reply astropy/stats/sigma_clipping.py]: done

File: astropy/stats/sigma_clipping.py (hunk 4 above)
Comment (larrybradley) [reply astropy/stats/sigma_clipping.py]: done

File: astropy/stats/tests/test_sigma_clipping.py (hunk 6 above)
Comment (larrybradley) [reply astropy/stats/tests/test_sigma_clipping.py]: fixed

File: astropy/stats/tests/test_sigma_clipping.py (hunk 7 above)
Comment (larrybradley) [reply astropy/stats/tests/test_sigma_clipping.py]: fixed

File: astropy/stats/sigma_clipping.py (hunk 3 above)
Comment (larrybradley) [reply astropy/stats/sigma_clipping.py:596]: indeed!

File: astropy/stats/sigma_clipping.py (hunk 5 above)
Comment (larrybradley) [reply astropy/stats/sigma_clipping.py:679]: I did some new profiling, and it's a bit hard to summarize in a single sentence.  Sometimes astropy is *faster* than scipy:

<pre>
size     % slower
1e8.     -6.9
1e7.     11.5
1e6.      6.3
1e5.     36.1
1e4.     10.6
1e3.    -18.2
</pre>

Perhaps we should just remove the part about the performance, and instead simply include the equivalent keywords to match `scipy.stats.sigmaclip`?

**Comment by larrybradley** [comment]
I just pushed another commit that fixes https://github.com/astropy/astropy/issues/11850 too.

File: astropy/stats/sigma_clipping.py (hunk 5 above)
Comment (mhvk) [reply astropy/stats/sigma_clipping.py:684]: Yes, sounds good! Though fine to do in a separate PR, if you prefer.

Comment (larrybradley) [reply astropy/stats/sigma_clipping.py]: I went ahead and did it here.  Thanks.

**Comment by larrybradley** [comment]
Ping @mhvk for final review.

**Review by mhvk — APPROVED** [review]
Yes, looks all good!

**Pull request merged**